identity-mdoc/src/commonMain/kotlin/com/android/identity/mdoc/connectionmethod/ConnectionMethodNfc.kt

```diff
@@ -70,5 +132,52 @@ class ConnectionMethodNfc(
                 map[OPTION_KEY_RESPONSE_DATA_FIELD_MAX_LENGTH].asNumber
             )
         }
+
+        internal fun fromNdefRecord(
+            record: NdefRecord,
+            role: MdocTransport.Role,
+        ): ConnectionMethodNfc? {
+            val payload = Buffer()
+            payload.write(record.payload)
+            val version = payload.readByte().toInt()
+            if (version != 0x01) {
+                Logger.w(TAG, "Expected version 0x01, found $version")
+                return null
+            }
+            val cmdLen = payload.readByte().toInt().and(0xff)
+            val cmdType = payload.readByte().toInt().and(0xff)
+            if (cmdType != 0x01) {
+                Logger.w(TAG, "expected type 0x01, found $cmdType")
+                return null
+            }
+            if (cmdLen < 2 || cmdLen > 3) {
+                Logger.w(TAG, "expected cmdLen in range 2-3, got $cmdLen")
+                return null
+            }
+            var commandDataFieldMaxLength = 0
+            for (n in 0 until cmdLen - 1) {
```

Comment:
If this is intended to read 1 byte per cmdLen, "until" already omits the end of the range. "0 until 2" provides 0 and 1. So "0 until 2 - 1" only provides 0, and we aren't reading all the bytes.

Same for rspLen, below.

Do we have unit tests to validate that we're reading this data correctly?

Reply:
This is WAI, the length includes `cmdType` (0x01 and 0x02). Added constants for those numbers. Added unit tests.